oak/server/rust/oak_runtime/src/node/roughtime.rs

```diff
@@ -16,32 +16,171 @@
+        let timeout_seconds = if config.timeout_seconds == 0 {
+            DEFAULT_TIMEOUT_SECONDS
+        } else {
+            config.timeout_seconds as u64
+        };
+        let server_retries = if config.server_retries == 0 {
+            DEFAULT_SERVER_RETRIES
+        } else {
+            config.server_retries as usize
+        };
+        let min_overlapping_intervals = if config.min_overlapping_intervals == 0 {
+            DEFAULT_MIN_OVERLAPPING_INTERVALS
+        } else {
+            config.min_overlapping_intervals as usize
+        };
+        let max_radius_microseconds = if config.max_radius_microseconds == 0 {
+            DEFAULT_MAX_RADIUS_MICROSECONDS
+        } else {
+            config.max_radius_microseconds
+        };
+        let servers = if config.servers.is_empty() {
+            get_default_servers()
+        } else {
+            config
+                .servers
+                .iter()
+                .map(|server| {
+                    RoughtimeServer::new(
+                        &server.name,
+                        &server.host,
+                        server.port as u16,
+                        &server.public_key_base64,
+                    )
+                })
+                .collect()
+        };
+
         Self {
             node_name: node_name.to_string(),
+            client: RoughtimeClient::new(
+                servers,
+                min_overlapping_intervals,
+                max_radius_microseconds,
+                timeout_seconds,
+                server_retries,
+            ),
         }
     }
+
+    /// Processes a gRPC invocation.
+    fn process_invocation(&self, runtime: &RuntimeProxy, invocation: &Invocation) {
+        match invocation.receive_request(runtime) {
+            Ok(request) => {
+                if request.method_name != "/oak.roughtime.RoughtimeService/GetRoughtime" {
+                    let message = format!("Unknwon method_name: {}", request.method_name);
+                    invocation.send_error(Code::NotFound, &message, runtime);
+                }
+                if RoughtimeRequest::decode(request.req_msg.as_slice()).is_err() {
+                    invocation.send_error(
+                        Code::InvalidArgument,
+                        "Could not parse request.",
+                        runtime,
+                    );
+                }
+            }
+            Err(error) => {
+                let message = format!("Error reading request: {:?}", error);
+                invocation.send_error(Code::InvalidArgument, &message, runtime);
+            }
+        };
+
+        match self.client.get_roughtime() {
+            Ok(time) => {
+                let reply = RoughtimeResponse {
+                    roughtime_usec: time,
+                };
+                let mut message = Vec::new();
+                match reply.encode(&mut message) {
+                    Ok(_) => {
+                        let response = GrpcResponse {
```

Comment:
nit: having both `reply` and `response` is a bit confusing; maybe `rsp` and `grpc_rsp` or somesuch?

Reply:
Renamed.